UI updates

levitte opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] CLA is signed

##### Description of change
<!-- Provide a description of the changes.

If it fixes a github issue, add Fixes #XXXX.
-->

- constify the input parameter to UI method getters.  *This is ABI compatible, it merely makes an already existing practice explicit*
- add space for an application data field in the UI structure, and functionality to access it.
- add a UI_METHOD wrapper for the PEM password callback, for applications that need to pass the latter to an interface that takes a UI_METHOD.
- document UI utility functions

levitte committed:
Enable apps to get a UI_METHOD for the default prompter

levitte committed:
Constify the input parameter to UI_method_get_*

levitte committed:
Add an application data field in the UI_METHOD

levitte committed:
Add a UI utility function with which to wrap pem_callback_cb in a UI_METHOD

levitte committed:
make update

levitte committed:
Document the UI utility functions

richsalz reviewed (changes requested):
any reason why not to use the standard EXDATA ?

levitte commented:
> any reason why not to use the standard EXDATA ?

not really, just that it escaped my mind.  Will fix

kaduk commented on doc/man3/UI_UTIL_read_pw.pod:
```diff
@@ -0,0 +1,47 @@
+UI_UTIL_read_pw_string() asks for a passphrase, using B<prompt> as a
+prompt, and stores it in B<buf>.
+The maximum allowed size is given with B<length>, including the ending
+NUL byte.
```
Maybe "trailing NUL", "NUL terminator", or "terminating NUL byte"?  No real preference from me, though.

levitte replied:
Thanks. Will change

kroeckx commented:
Please propose some policy about which changes are allowed in public headers.

kroeckx commented:
Is there a reason this isn't tested?

levitte commented:
> Please propose some policy about which changes are allowed in public headers.

Do these suffice? They say nothing about headers and everything about backward ABI compatibility. 

https://www.openssl.org/docs/faq.html#MISC8
https://www.openssl.org/policies/releasestrat.html

The assumption is that as long as the API doesn't break applications and ABI compatibility is maintained, we should be fine.

levitte commented:
> Is there a reason this isn't tested?

I'm writing a test, will show up in a commit in a couple of hours.

levitte committed:
[to be squashed] libcrypto.num fixup

levitte committed:
[to be squashed] make the ui_method appdata an ex_data

levitte committed:
[to be squashed] make the pem callback wrapper use the ui_method ex_data

levitte committed:
Add a few documentation lines about UI_OpenSSL()

levitte committed:
Add a test "uitest"

It tests both the use of UI_METHOD (through the apps/apps.h API) and
wrapping an older style PEM password callback in a UI_METHOD.

Replace the earlier UI test with a run of this test program

levitte committed:
[to be squashed] doc update of UI utility functions

levitte commented:
I've made a number of updates:

- changed the UI_METHOD app data to be an ex_data field (ping @richsalz)
- added a test program that tests the UI_METHOD wrapper as well as the method already tested through the `openssl` command.  This simplified the UI test recipe.

richsalz reviewed (approved):
Looks okay to me.

levitte committed:
[to be squashed] UI util: rename "index" to "idx"

The compiler complained that it shadows index() from strings.h

levitte committed:
[to be squashed] uitest: fix memory leak

levitte commented:
Pushed a couple of fixes for the travis build issues

levitte commented:
Builds are passing now.  @richsalz, would you mind reconfirming?

richsalz commented:
sure, +1

levitte commented:
Perfect.  Thanks.  Merged
